src/cmd/build.js

```diff
@@ -30,31 +33,48 @@ export default function build(
     fileFilter = new FileFilter();
   }
 
-  return resolveManifest
-    .then((manifestData) =>
-      Promise.all([
-        prepareArtifactsDir(artifactsDir),
-        zipDir(sourceDir, {
-          filter: (...args) => fileFilter.wantFile(...args),
-        }),
-      ])
-      .then((results) => {
-        let [artifactsDir, buffer] = results;
-        let packageName = safeFileName(
-          `${manifestData.name}-${manifestData.version}.xpi`);
-        let extensionPath = path.join(artifactsDir, packageName);
-        let stream = createWriteStream(extensionPath);
-        let promisedStream = streamToPromise(stream);
-
-        stream.write(buffer, () => stream.end());
-
-        return promisedStream
-          .then(() => {
-            log.info(`Your web extension is ready: ${extensionPath}`);
-            return {extensionPath};
-          });
-      })
-    );
+  if (!createPackage) {
+    createPackage = () => {
+      return resolveManifest
+        .then((manifestData) => {
+          return zipDir(
+            sourceDir, {
+              filter: (...args) => fileFilter.wantFile(...args),
+            })
+            .then((buffer) => {
+              let packageName = safeFileName(
+                `${manifestData.name}-${manifestData.version}.xpi`);
+              let extensionPath = path.join(artifactsDir, packageName);
+              let stream = createWriteStream(extensionPath);
+              let promisedStream = streamToPromise(stream);
+
+              stream.write(buffer, () => stream.end());
+
+              return promisedStream
+                .then(() => {
+                  log.info(`Your web extension is ready: ${extensionPath}`);
+                  return {extensionPath};
+                });
+            });
+        });
+    };
+  }
+
+  return prepareArtifactsDir(artifactsDir)
+    .then(() => createPackage())
+    .then((result) => {
+      if (rebuildWhenSourceChanges) {
+        log.info('Rebuilding when files change...');
+        onSourceChange({
+          sourceDir, artifactsDir, onChange: () => {
+            return createPackage().catch((error) => {
+              throw error;
```

Comment:
This error is not currently logged (it probably ends up into the node [uncaughtexception](https://nodejs.org/api/process.html#process_event_uncaughtexception).

we risk to don't warn the user of errors that he probably wants to know about, e.g.:
- if we during a "web-ext build --as-needed" session, we suddently remove the `web-ext-artifacts` dir, currently no error is shown in the web-ext logs.

we can even just log the error in addition to the `throw` (maybe with the stack trace information when it is running in verbose mode), e.g. something like:
```js
return createPackage().catch((error) => {
    log.error(`Error during build: ${error}`);
    log.debug(`Last error debug info: ${error.stack}`);
    throw error;
}
```